blscurve/blst/blst_min_pubkey_sig_core.nim

```diff
@@ -378,3 +381,157 @@ func finish*(ctx: var ContextCoreAggregateVerify, signature: Signature or Aggreg
 
   ctx.commit()
   result = bool ctx.finalVerify()
+
+# Parallelized Batch Verifier primitives
+# ----------------------------------------------------------------------
+#
+# Ultimately this can be merged with the internal ContextCoreAggregateVerify
+# but:
+# - The previous code was audited
+# - For now we only support BLST
+#   though MIRACL is straightforward, context merge is just a FP12 multiplication
+#   since GT is a multiplicative group.
+# - we need to hold cryptographycally secure
+#   random bytes to protect against forged signatures
+#   that would not pass non-aggregated verification
+#   via random scalar blinding
+#   https://ethresear.ch/t/fast-verification-of-multiple-bls-signatures/5407/14
+#
+#
+# Tradeoffs:
+# - Adds:
+#   - a blinding multiplication on G2
+#   - a blinding mul on G1 or Fp
+# - Saves:
+#   - Only 1 Miller Loop per thread (50% pairing)
+#   - Only 1 Final Exponentiation per batch (50% pairing cost)
+#
+# Assuming blinding muls cost 60% of a pairing (worst case with 255-bit blinding)
+# verifying 3 signatures would have a base cost of 300
+# Batched single threaded the cost would be
+# 60*3 (blinding 255-bit) + 50 (Miller) + 50 (final exp) = 280
+#
+# With 64-bit blinding and ~20% overhead
+# (not 15% because no endomorphism acceleration with 64-bit)
+# 20*3 (blinding 64-bit) + 50 (Miller) + 50 (final exp) = 160
+#
+# If split on 2 cores, the critical path is
+# 20*2 (blinding 64-bit) + 50 (Miller) + 50 (final exp) = 140
+#
+# If split on 3 cores, the critical path is
+# 20*1 (blinding 64-bit) + 50 (Miller) + 50 (final exp) = 120
+
+type
+  ContextMultiAggregateVerify*[DomainSepTag: static string] = object
```

Comment:
why is this generic? it looks like something that's dynamically passed to the library at init time

Reply:
The domain separation tag can only have 3 relevant values from the BLS spec https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-02#section-4.2.1:
- `BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_`
- `BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_AUG_`
- `BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_` (Eth 2 spec)

Since we focus on Eth2, the constant is hardcoded in the public API and static string in the internal (core) implementation.
https://github.com/status-im/nim-blscurve/blob/31ada540789fc0081b4fcfd6f693410683235fd7/blscurve/bls_sig_min_pubkey.nim#L15-L32

This has a couple of advantages:
- no need to carry it in a field and so have string allocation
- no need to change the hashToG2 procedures because at a low-level hashToG2 needs to append a fixed byte to the input https://github.com/status-im/nim-blscurve/blob/31ada540789fc0081b4fcfd6f693410683235fd7/blscurve/miracl/hash_to_curve.nim#L46-L60
  if we do it at compile-time we don't need to reallocate and ensure no GC operation in low-level crypto primitives.